python/plugins/processing/algs/qgis/Climb.py

```diff
@@ -0,0 +1,255 @@
+
+        # get features from source (with z values)
+        features = layerwithz.getFeatures()
+        totalclimb = 0
+        totaldescent = 0
+        minelevation = float('Infinity')
+        maxelevation = float('-Infinity')
+
+        no_z_nodes = []
+
+        for current, feature in enumerate(features):
+            # Stop the algorithm if cancelled
+            if feedback.isCanceled():
+                break
+            climb = 0
+            descent = 0
+            minelev = float('Infinity')
+            maxelev = float('-Infinity')
+            # In case of multigeometries we need to do the parts
+            parts = feature.geometry().constParts()
+            for part in parts:
+                # Calculate the climb
+                first = True
+                zval = 0
+                for idx, v in enumerate(part.vertices()):
+                    zval = v.z()
+                    if math.isnan(zval):
+                        no_z_nodes.append(idx)
+                        continue
+                    if first:
+                        prevz = zval
+                        minelev = zval
+                        maxelev = zval
+                        first = False
+                    else:
+                        diff = zval - prevz
+                        if diff > 0:
+                            climb = climb + diff
+                        else:
+                            descent = descent - diff
+                        if minelev > zval:
+                            minelev = zval
+                        if maxelev < zval:
+                            maxelev = zval
+                    prevz = zval
+                totalclimb = totalclimb + climb
+                totaldescent = totaldescent + descent
+            # Set the attribute values
+            attrs = feature.attributes()
+            outattrs = []
+            attrindex = 0
+            for attr in attrs:
+                # Skip attributes from the input layer that had names
+                # that were equal to the generated ones
+                if not (attrindex == climbindex or
+                        attrindex == descentindex):
+                    outattrs.append(attr)
+                attrindex = attrindex + 1
+            # feature.setAttributes(outattrs + [climb, descent])
+            feature.setAttributes(outattrs +
+                                  [climb, descent, minelev, maxelev])
+            # Add a feature to the sink
+            sink.addFeature(feature, QgsFeatureSink.FastInsert)
+            if minelevation > minelev:
+                    minelevation = minelev
+            if maxelevation < maxelev:
+                    maxelevation = maxelev
+            # Update the progress bar
+            if fcount > 0:
+                feedback.setProgress(int(100 * current / fcount))
+
+        feedback.pushInfo(self.tr('Following nodes miss Z value ') + str(no_z_nodes))
```

Comment:
Since this reports the point number within a single line, it should be moved to line 226 to have a meaning.  The alternative is to include feature id and part number.
I also suggest to change the text to:
'The following points on the line do not have Z values: '

Reply:
@havatv I think Matthias suggested to grab all the no Z nodes indexes and put them into a string to show at the end of the algorithm, else with a string with a lot of nodes it could be a little bit overkilling to push information each time. The approach of the next comment (with the enhancements to what the string should contain) is maybe better?

Reply:
Fine with me. I have added some suggestions to allow that.

Reply:
Can you switch to template based style here too